cinnabar/wrangle.py

```diff
@@ -22,209 +23,137 @@ def read_csv(filepath: pathlib.Path) -> dict:
+    >>> experimental_result2 = ExperimentalResult("CAT-17g", expt_DG=-9.73, expt_dDG=0.10)
+    >>> # Load/create calculated results
+    >>> calculated_result = RelativeMeasurement("CAT-13a", "CAT-17g", calc_DDG=0.36, mbar_error=0.11, other_error=0.0)
+    >>> # Incrementally created FEMap
+    >>> fe = FEMap()
+    >>> fe.add_measurement(experimental_result1)
+    >>> fe.add_measurement(experimental_result2)
+    >>> fe.add_measurement(calculated_result)
+    """
+    # MultiGraph allows multiple *relative* measurements, but not multiple identical nodes
+    # so maybe instead make absolute results all relative to a dummy node
+    computational_graph: nx.MultiDiGraph
+    experimental_graph: nx.MultiDiGraph
 
-        self.generate_graph_from_results()
+    @classmethod
+    def from_csv(cls, filename):
+        """Construct from legacy csv format"""
+        data = read_csv(filename)
 
-    def generate_graph_from_results(self):
-        self._name_to_id = {}
-        self._id_to_name = {}
-        id = 0
-        for result in self.results["Calculated"]:
-            if result.ligandA not in self._name_to_id.keys():
-                self._name_to_id[result.ligandA] = id
-                self._id_to_name[id] = result.ligandA
-                id += 1
-            if result.ligandB not in self._name_to_id.keys():
-                self._name_to_id[result.ligandB] = id
-                self._id_to_name[id] = result.ligandB
-                id += 1
-            # # TODO need some exp error for mle to converge for exp... this is a horrible hack
-            self.graph.add_edge(
-                self._name_to_id[result.ligandA],
-                self._name_to_id[result.ligandB],
-                calc_DDG=result.calc_DDG,
-                calc_dDDG=result.calc_dDDG,
-            )
-
-        for node in self.graph.nodes(data=True):
-            name = self._id_to_name[node[0]]
-            node[1]["name"] = name
-            node[1]["exp_DG"] = self.results["Experimental"][name].DG
-            node[1]["exp_dDG"] = self.results["Experimental"][name].dDG
-
-        for edge in self.graph.edges(data=True):
-            DG_A = self.graph.nodes[edge[0]]["exp_DG"]
-            DG_B = self.graph.nodes[edge[1]]["exp_DG"]
-            edge[2]["exp_DDG"] = DG_B - DG_A
-            dDG_A = self.graph.nodes[edge[0]]["exp_dDG"]
-            dDG_B = self.graph.nodes[edge[1]]["exp_dDG"]
-            edge[2]["exp_dDDG"] = (dDG_A**2 + dDG_B**2) ** 0.5
-
-        self.n_ligands = self.graph.number_of_nodes()
-        self.n_edges = self.graph.number_of_edges()
-        self.degree = self.n_edges / self.n_ligands
-
-        # check the graph has minimal connectivity
-        self.check_weakly_connected()
-        if not self.weakly_connected:
-            print("Graph is not connected enough to compute absolute values")
+        # unpack data dictionary
+        fe = cls()
+        for r in data['Calculated']:
```

Comment:
to save my sanity, do we really need to have capitalization here and for "Experimental"?

Reply:
this is legacy stuff, I'm not improving it (and it'll get removed eventually)

Reply:
I thought we were boldly changing stuff, but yeah fair I can't wait :P